File: app/src/main/java/me/zhanghai/android/douya/network/api/ApiAuthenticator.java
/*
 * Copyright (c) 2017 Zhang Hai <Dreaming.in.Code.ZH@Gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.douya.network.api;

import android.accounts.Account;

import java.io.IOException;

import me.zhanghai.android.douya.network.AndroidAuthenticator;
import okhttp3.Response;
import okhttp3.ResponseBody;

public class ApiAuthenticator extends AndroidAuthenticator {

    public ApiAuthenticator(Account account, String authTokenType) {
        super(account, authTokenType, true);
    }

    @Override
    public boolean shouldRetryAuthentication(Response response) {
        // Don't consume the response body so that it can later be read by others.
        ResponseBody responseBody;
        try {
            responseBody = response.peekBody(Long.MAX_VALUE);
        } catch (IOException e) {
            e.printStackTrace();
            return false;
        }
        //noinspection ThrowableInstanceNeverThrown
        ApiError apiError = new ApiError(response, responseBody);
        switch (apiError.code) {
            case ApiContract.Response.Error.Codes.Token.INVALID_ACCESS_TOKEN:
            case ApiContract.Response.Error.Codes.Token.ACCESS_TOKEN_HAS_EXPIRED:
            case ApiContract.Response.Error.Codes.Token.INVALID_REFRESH_TOKEN:
            case ApiContract.Response.Error.Codes.Token
                    .ACCESS_TOKEN_HAS_EXPIRED_SINCE_PASSWORD_CHANGED:
                return true;
            default:
                return false;
        }
    }
}

File: app/src/main/java/me/zhanghai/android/douya/api/app/ApiAuthenticator.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.douya.api.app

import android.accounts.Account
import me.zhanghai.android.douya.api.info.ApiContract
import me.zhanghai.android.douya.network.AndroidAuthenticator
import okhttp3.Response
import timber.log.Timber
import java.io.IOException

class ApiAuthenticator(account: Account) : AndroidAuthenticator(account) {
    override fun shouldRetryAuthentication(response: Response): Boolean {
        // Don't consume the response body so that it can later be read by others.
        val body = try {
            response.peekBody(Long.MAX_VALUE)
        } catch (e: IOException) {
            Timber.e(e)
            return false
        }
        return when (body.toErrorResponse()?.code) {
            ApiContract.Error.Codes.INVALID_ACCESS_TOKEN,
            ApiContract.Error.Codes.ACCESS_TOKEN_HAS_EXPIRED,
            ApiContract.Error.Codes.ACCESS_TOKEN_HAS_EXPIRED_SINCE_PASSWORD_CHANGED -> true
            else -> false
        }
    }
}
